### ISSUE TITLE
Does it work with MMTeX?

### ISSUE THREAD

**Issue by ghost** [vlasakm/optex-minim#1]
Hi! I've tried compiling a copy of `optex-minim-mp-test.tex` using this package and  [MMTeX](https://github.com/vlasakm/mmtex), but it complains about `\newcount` being undefined. The log is quite long, but these are the first lines, being the next ones similar:
```
! Undefined control sequence.
<recently read> \newcount 
          
<to be read again> 
\e@alloc@attribute@count 
\alloc:makenew ... \newcount \csname #3\endcsname 
                                                  \csname #3\endcsname 0\fi ...

l.141 ... \attributedef    {e@alloc@attribute@count}
                                                  
! Undefined control sequence.
<recently read> \newcount 
          
<to be read again> 
\e@alloc@whatsit@count 
\alloc:makenew ... \newcount \csname #3\endcsname 
                                                  \csname #3\endcsname 0\fi ...

l.142 ...   \chardef         {e@alloc@whatsit@count}
                                                  
! Undefined control sequence.
<recently read> \newcount 
          
<to be read again> 
\e@alloc@bytecode@count 
\alloc:makenew ... \newcount \csname #3\endcsname 
                                                  \csname #3\endcsname 0\fi ...


```
I've added files from `minim`, as well as `ltluatex.lua`, `ltluatex.lua`, and `etex.src`, to `ls-R`, so files seem to be found. What could be happening? IMHO a self-contained `optex-minim` would be easier to handle with.

**Comment by vlasakm** [comment]
Hello! You have stumbled on "in development stuff". The goal is that it indeed should work in MMTeX (and TeX Live and MikTeX, etc.), but the process of getting minim and OpTeX work is not yet finished.

Just yesterday Esger Renkema (of minim) pushed changes to the `master` branch, which makes it work with my `newresources` branch of OpTeX and `master` branch of `optex-minim`.

So currently, you would need to get this to try the examples:

https://github.com/vlasakm/optex-minim/tree/master
https://gitlab.com/renkema/minim/-/tree/master
https://github.com/vlasakm/OpTeX/tree/newresources

Since it is in IMO finished form, hopefully we can get all three released on CTAN in the next few days.

I plan to include minim, minim-optex and PGF in the next release of MMTeX, since the next OpTeX version hopefully makes them all usable.

> I've added files from minim, as well as ltluatex.lua, ltluatex.lua, and etex.src, to ls-R

Adding `ltluatex` and `etex` shouldn't be necessary. As I said the development versions of all three packages are needed.

Also, the documentation of mmtex is still lacking since I deleted the obsolete documentation (https://github.com/vlasakm/mmtex/releases/download/v20201030/mmtex-doc.pdf). Most of it is really obsolete, but section 12 about the file searching is still relevant and is the way to add new files to mmtex. Depending on how structured is the directory I want to add files from, I add an entry either to `TEXMFDOTDIR` (for flat directory) or `TEXMF` (for texmf trees, where TeX files are in `tex` subdirectory, fonts in `fonts`, etc.

My `~/.config/mmtex/temxf.cnf` file as of now looks like this:

```
error_line      =     250
half_error_line =     125
max_print_line  =  100000

%TEXMF = ~/.local/share/mmtex
TEXMF = ~/src/pgf
TEXMFDOTDIR = .:~/src/CTUstyle3:~/projects/pdfextra{/examples,/extref,}:~/projects/optex-minim/:~/src/optex/optex/base:~/src/optex/optex/doc:~/src/optex/optex/pkg:~/src/minim//
```
I don't like TeX console output wrapped to 78 charactes, so I set the limit much higher. For optex-minim I needed to test PGF/TikZ which has texmf structure and I have it cloned in `~/src/pgf`, but I normally install the [`".tds.zip"`](https://mirrors.ctan.org/install/graphics/pgf/base/pgf.tds.zip) files of some CTAN packages to `~/.local/share/mmtex`. I also use development versions of minim and OpTeX - these have mostly flat structure, so I add them to `TEXMFDOTDIR` (but minim has multiple of these flat directories - instead of naming them all I just recursively added all subdirectories with trailing `//`.

> IMHO a self-contained optex-minim would be easier to handle with.

Do you mean including all of minim in optex-minim or something more?

My thinking was, that standalone package doesn't cause any duplication with minim, and upstream updates to minim from Esger Renkema will automatically propage. When optex-minim is released on CTAN, everyone should be able to just do `\load[minim-mp]` (or `minim-pdf`) in their TeX distribution of choice (whether it is MMTeX or TeX Live).

**Comment by vlasakm** [comment]
Did you install mmtex from source? Would a preview version of MMTeX with development versions of minim/optex/optex-minim be useful to you now?

**Comment by ghost** [comment]
> So currently, you would need to get this to try the examples:
> 
> https://github.com/vlasakm/optex-minim/tree/master https://gitlab.com/renkema/minim/-/tree/master https://github.com/vlasakm/OpTeX/tree/newresources

I made some replacements and now it works. Thanks for the pointers.
 
> I plan to include minim, minim-optex and PGF in the next release of MMTeX, since the next OpTeX version hopefully makes them all usable.

Cool

> Do you mean including all of minim in optex-minim or something more?

I'm rather interested in MetaPost, as OpTeX doesn't support it by default and I don't use TikZ at all. I realized that either luatex-plain files (from ConTeXt) or luamplib (wtih some patches) make MetaPost usable in OpTeX. I'm happy to see minim is also working now.

**Comment by ghost** [comment]
> Did you install mmtex from source?

Yes. It's very, very customizable and it makes easy to update or remove dependencies (e.g. I removed SyncTeX and updated Lua to 5.4 with a few patches). Also, I don't use TeX Live, only standalone distributions (ConTeXt and, now, MMTeX).

> Would a preview version of MMTeX with development versions of minim/optex/optex-minim be useful to you now?

Yes, but I can wait for updates. Thank you a lot for your answer and MMTeX.

**Comment by vlasakm** [comment]
I am curious about Lua 5.4, do you have any need for it? Out of curiousity, what did you patch? When I experimented with LuaTeX sources (to try to strip it from external components), I also tried upgrading it to Lua 5.3 / 5.4 style, e.g.:

https://github.com/vlasakm/luatex-experiments/commit/0b773ccab66e5f70d72c8585afe3a00848cd21ba
https://github.com/vlasakm/luatex-experiments/commit/c9cc58529c8a88a6875013d73a5269991d66feb3

Would it help you if I upgraded to Lua 5.4, perhaps by using your patches? Though I only did my 5.4 update on stripped down version of LuaTeX, so I am not sure if there wouldn't be incompatibilities with `luaffi` and such. Do you use that?

Also in the case of SyncTeX - I also have no use for it. And as ConTeXt demonstrates a custom output is possible with LuaTeX, so maybe it is not needed after all, and there would be one less external thing in mmtex.

> > Would a preview version of MMTeX with development versions of minim/optex/optex-minim be useful to you now?
> 
> Yes, but I can wait for updates. Thank you a lot for your answer and MMTeX.

It seems that minim is near a release. If you can wait a bit I will wait for everything to release properly on CTAN.

**Comment by ghost** [comment]
> I am curious about Lua 5.4, do you have any need for it? Out of curiousity, what did you patch? When I experimented with LuaTeX sources (to try to strip it from external components), I also tried upgrading it to Lua 5.3 / 5.4 style, e.g.:

Mainly files in `lua/src`, which have been patched in LuaTeX to support deprecated functions of Lua 5.1 and Lua 5.2 needed by `luasocket`. Also, I replaced `texluac.c` by Lua 5.4's `luac.c`, but maybe I'm missing something else here as `texluac.c` seems to be tuned for LuaTeX. I'll make an exhaustive list later.

> [vlasakm/luatex-experiments@0b773cc](https://github.com/vlasakm/luatex-experiments/commit/0b773ccab66e5f70d72c8585afe3a00848cd21ba) [vlasakm/luatex-experiments@c9cc585](https://github.com/vlasakm/luatex-experiments/commit/c9cc58529c8a88a6875013d73a5269991d66feb3)
> 
> Would it help you if I upgraded to Lua 5.4, perhaps by using your patches? Though I only did my 5.4 update on stripped down version of LuaTeX, so I am not sure if there wouldn't be incompatibilities with `luaffi` and such. Do you use that?
> 

I'd have to check my files, but it was rather toying with LuaTeX in order to know how far can I go with removing unused dependencies and adding my own extensions. As I don't really understand how Git works, I'm making changes to a local copy according to TeX Live patches. I forgot to mention I don't use ffi either and I have removed it, too. I'll try to make something more ordered.

**Comment by vlasakm** [comment]
> I'd have to check my files, but it was rather toying with LuaTeX in order to know how far can I go with removing unused dependencies and adding my own extensions.

Very interesting indeed! The removing dependencies part is what I also explored. https://github.com/vlasakm/luatex-experiments/tree/master. Though I haven't yet followed up on replacing parts that I removed (kpathsea, PDF creation, processing of images).

**Comment by vlasakm** [comment]
New version of minim and OpTeX (with the minim support included) have been released, so I also released a new version  of mmtex which has that and more.

Let me know if it works for you.
